logger.js

```diff
@@ -85,58 +85,56 @@ function pinoLogger (opts, stream) {
   delete opts.useLevel
 
   const genReqId = reqIdGenFactory(opts.genReqId)
-  loggingMiddleware.logger = logger
-  return loggingMiddleware
-
-  function onResFinished (err) {
-    this.removeListener('close', onResFinished)
-    this.removeListener('error', onResFinished)
-    this.removeListener('finish', onResFinished)
+  const result = (req, res, next) => {
+    return loggingMiddleware(logger, req, res, next)
+  }
+  result.logger = logger
+  return result
 
-    let log = this.log
-    const responseTime = Date.now() - this[startTime]
-    const level = getLogLevelFromCustomLogLevel(customLogLevel, useLevel, this, err)
+  function onResFinished (res, logger, err) {
+    let log = logger
+    const responseTime = Date.now() - res[startTime]
+    const level = getLogLevelFromCustomLogLevel(customLogLevel, useLevel, res, err)
 
     if (level === 'silent') {
       return
     }
 
-    const req = this[reqObject]
-    const res = this
+    const req = res[reqObject]
 
     const customPropBindings = (typeof customProps === 'function') ? customProps(req, res) : customProps
     if (customPropBindings) {
-      log = this.log.child(customPropBindings)
+      log = logger.child(customPropBindings)
     }
 
-    if (err || this.err || this.statusCode >= 500) {
-      const error = err || this.err || new Error('failed with status code ' + this.statusCode)
+    if (err || res.err || res.statusCode >= 500) {
+      const error = err || res.err || new Error('failed with status code ' + res.statusCode)
 
       log[level](
```

Comment:
the current implementation emits the final log message on the single logger attached to the response object. Instead we have passed the intended logger in as a parameter from a new closure onResponseComplete so that each middleware generates it's logs to its own pino logger, not relying on the logger attached to the response object.